src/ips.rs

```diff
@@ -0,0 +1,523 @@
+use spectre::edge::Edge;
+use spectre::graph::{AGraph, Graph};
+use std::collections::HashMap;
+use std::collections::{HashSet, VecDeque};
+use std::net::IpAddr;
+use std::str::FromStr;
+
+/// Intelligent Peer Sharing (IPS) module structure
+#[derive(Default, Clone)]
+pub struct Ips {
+    config: IPSConfiguration,
+    degree_factors: NormalizationFactors,
+    betweenness_factors: NormalizationFactors,
+    closeness_factors: NormalizationFactors,
+    eigenvector_factors: NormalizationFactors,
+}
+
+/// Peer list structure containing peer list for each node
+#[derive(Clone, Serialize, Deserialize)]
+pub struct Peer {
+    /// IP address of the node
+    pub ip: IpAddr,
+    /// List of peers for the node
+    pub list: Vec<IpAddr>,
+}
+
+const NORMALIZE_TO_VALUE: f64 = 100.0;
+#[derive(Default, Clone)]
+struct NormalizationFactors {
+    min: f64,
+    max: f64,
+}
+
+impl Ips {
+    pub fn new(config: IPSConfiguration) -> Ips {
+        Ips {
+            config,
+            ..Default::default()
+        }
+    }
+
+    /// Generate peer list - main function with The Algorithm
+    /// It needs state and agraph to be passed as parameters which need to be correlated with
+    /// the crawler's state and agraph (and with each other), so the indexes saved in the
+    /// agraph are the same as the positions of the nodes in the state.nodes.
+    pub async fn generate(&mut self, state: &CrunchyState, agraph: &AGraph) -> Vec<Peer> {
+        let mut peer_list = Vec::new();
+
+        // Reconstruct graph from the agraph - we need to do this because we need all the
+        // measures provided by spectre's graph implementation.
+        // Using agraph gives us certainity that we are using the same graph as the crawler and
+        // there are only good nodes there (this is critical assumption!). Second assumption is
+        // that agraph node indexes are the same as in the state.nodes vector.
+        let mut graph = self.construct_graph(&state.nodes, agraph);
+
+        // 0 - Detect islands
+        // To reconsider if islands should be merged prior to any other computations or not.
+        // IMHO, if there are islands they can influence on the results of the computations.
+        // TODO(asmie): Merging islands is not implemented yet.
+        let _islands = self.detect_islands(agraph);
+
+        // Now take the current params
+        let degrees = graph.degree_centrality();
+        let degree_avg = self.degree_centrality_avg(&degrees);
+        let eigenvalues = graph.eigenvalue_centrality();
+
+        // Determine factors used for normalization.
+        // Normalization step is needed to make sure that all the factors are in the same range and
+        // weights can be applied to them.
+        self.determine_degrees_factors(&degrees);
+        self.determine_eigenvalue_factors(&eigenvalues);
+        self.determine_betweenness_factors(&state.nodes);
+        self.determine_closeness_factors(&state.nodes);
```

Comment:
It seems like these four functions are calculating min/max for a given set of numbers (which can have a different type, but all implement [PartialOrd](https://doc.rust-lang.org/std/cmp/trait.PartialOrd.html#))

One alternative we could consider could be a generic function with a generic parameter (I'll throw multiple ideas in the example)
```rust
let betweenness = [...]; // collected values from state.node.iter() for betweenness
let closeness = [...]; // collected values from state.node.iter() for closeness

self.degrees_factors = NormalizationFactors::determine(&deegrees.values().cloned().collect());
self.eigenvector_factors = NormalizationFactors::determine(&eigenvalues.values().cloned().collect());
self.betweenness_factors = NormalizationFactors::determine(&betweenness);
self.closeness_factors = NormalizationFactors::determine(&closeness);
```

Reply:
With some issues but done :)